MudStepper: New Component

fondraco opened the pull request:
## Description
I want to create a stepper component; I've done some base work to kick it off. It should resolve #881. If you have any ideas for functionality, code implementation or anything else, please speak your mind. I'm very open to it. I'd love to see the MudStepper happen.

![image](https://github.com/MudBlazor/MudBlazor/assets/44090/1dcae021-eee2-4eef-94e0-b8ced6258ea8)
![image](https://github.com/MudBlazor/MudBlazor/assets/44090/604b46c0-e338-41f9-b006-2fec22d0c96d)

## How Has This Been Tested?
This will need a few tests.

## Types of changes
<!-- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

![image](https://github.com/MudBlazor/MudBlazor/assets/49674671/af559cd6-79eb-410c-90e0-f578f97d6a6b)

## Checklist:
<!-- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] The PR is submitted to the correct branch (`dev`).
- [x] My code follows the code style of this project.
- [ ] I've added relevant tests.

fondraco committed:
MudStepper: Laid basis for the new Stepper component (#881)

fondraco committed:
MudStepper: Updated CSS for the component (#881)

ScarletKuro commented:
I see that you have logic in component's parameters, I would highly like to avoid that in new components. I know we have this pattern in MudBlazor code base, but it's incorrect, and in V7 there is a plan to eliminate all this places.
Currently we have the BL0007 warring disabled, but it fact it doesn't allow to write have the parameter property to be other than an auto property.
Please refer to this links for more details on why is this dangerous: https://github.com/dotnet/aspnetcore/issues/26230, https://github.com/dotnet/aspnetcore/issues/24599#issuecomment-697588562, https://github.com/dotnet/aspnetcore/issues/28518

fondraco commented:
> it doesn't permit the parameter property to be anything other than an auto property.

So, can the `ActiveIndex` be bound if the setter would need a method like this and the user can't just choose an arbitrary number?

```
private async void SetActiveIndex(int value)
    {
        var validPanel = _steps.Count > 0 && value != -1 && value <= _steps.Count - 1;

        if (ActiveIndex != value)
        {
            ActiveStep = validPanel ? _steps[value] : null;
            await ActiveIndexChanged.InvokeAsync(ActiveIndex = value);
        }
        else if (validPanel)
            ActiveStep = _steps[value];
    }
```

As I assume, the set should be private, and the public method should handle that.

ScarletKuro commented:
Here are some additional resources / issues I recommend for a better understanding of [Parameter]:
1. https://github.com/dotnet/aspnetcore/issues/45152
2. https://github.com/dotnet/aspnetcore/discussions/49187#discussioncomment-6372174
3. https://learn.microsoft.com/en-us/aspnet/core/blazor/components/overwriting-parameters?view=aspnetcore-7.0
4. https://learn.microsoft.com/en-us/aspnet/core/blazor/components/data-binding?view=aspnetcore-7.0#binding-with-component-parameters

> So, can the `ActiveIndex` be bound if the setter would need a method like this and the user can't just choose an arbitrary number?

If `ActiveIndex` cannot accept arbitrary numbers, there are two theoretical solutions to consider:
1. Instead of exposing `ActiveIndex` as a parameter, you can expose an `EventCallback` that notifies when the value changes. However, the value can only be modified through a public method like `SetActiveIndex`. This approach, although slightly less convenient as it requires the use of `ref` to access the method, is used by the [CodeBeam.MudBlazor.Extensions](https://github.com/CodeBeamOrg/CodeBeam.MudBlazor.Extensions/blob/28cc7f053df67905866303ae9e57446238cfd0fa/CodeBeam.MudBlazor.Extensions/Components/Stepper/MudStepper.razor.cs).
5. Another option is to expose both `ActiveIndex` as a component parameter and `ActiveIndexChanged` as an `EventCallback`, indicating that it should be bindable. When an outside component changes `ActiveIndex` to a value outside the acceptable range, you can override the `SetParametersAsync` method to validate the value and call `ActiveIndexChanged` with the corrected value. This change will trigger the parent component to refresh itself with the new value. For a concrete implementation example, you can refer to how [RadzenSteps](https://github.com/radzenhq/radzen-blazor/blob/fca72e3faded705b7201076e99ab6b917a23ebd5/Radzen.Blazor/RadzenSteps.razor.cs#L307C45-L307C45) handles this scenario. If you require the `DidParameterChange` functionality, please let me know, and I can submit it as a separate pull request (we plan to use it future, but with different name). Radzen provides a well-implemented example, except that the logic for `NextText`, `PreviousText`, and `SelectedIndex` should be moved to to `OnParametersSetAsync`, as suggested in https://github.com/dotnet/aspnetcore/issues/45152.

fondraco commented:
I'd prefer the second option. It seems like more work, but the result will pay off. Thank you for the articles; that set me on the right track.

It's still unclear to me whether the user should have a lot of freedom within steps movement. I've implemented for the movement the `ProcessStep` method (it may need some reworking) that handles the movement and as is in Tabs, I added the `Func OnPreviewInteraction` so the user can step in and alter the behaviour.

Is it a good idea to implement `Disabled` and `Completed` param for each step, so the user can alter that as well?

fondraco committed:
MudStepper: Updated logic and added alternate label visual (#881)

fondraco commented:
I added `AlternateLabel` as a new visual and also controls.

Basic horizontal stepper
![image](https://github.com/MudBlazor/MudBlazor/assets/49674671/5401a712-f04e-415d-9599-1aad3bb377c6)

Alternate label horizontal stepper
![image](https://github.com/MudBlazor/MudBlazor/assets/49674671/1decc290-2a2c-46c8-936a-9d5a40aac625)

fondraco committed:
MudStepper: Added visual option for Vertical stepper, added an example of this new feature (#881)

fondraco committed:
MudStepper: updated visuals for the component (#881)

fondraco commented:
I added a Vertical state for the component. By the way, those nav buttons on the end are not supposed to be there like that; that's only for testing purposes.
![image](https://github.com/MudBlazor/MudBlazor/assets/49674671/9138a888-e089-4fd9-b3bc-d4dd2ff7b056)

henon commented:
> It's still unclear to me whether the user should have a lot of freedom within steps movement.

If it can be avoided, don't restrict freedoms. Even if most won't need them, there are bound to be users who need them.

henon commented:
> Is it a good idea to implement `Disabled` and `Completed` param for each step, so the user can alter that as well?

Yes I think so. For instance, if in the middle the user of the stepper wants to go back to a previous step and there changes something that invalidates following steps which they had already completed then you'd need this, right?

henon commented on src/MudBlazor/Components/Stepper/MudStepper.razor.cs:
```diff
@@ -0,0 +1,241 @@
+    [Parameter]
+    public string StepClass { get; set; }
+    
+    /// <summary>
+    /// Renders labels for each step title below the circle
+    /// </summary>
+    [Parameter]
+    public bool AlternateLabel { get; set; }
```
Couldn't that be computed by looking at all the steps?

fondraco replied:
This enables the user to swap between visual options, as mentioned in [this comment](https://github.com/MudBlazor/MudBlazor/pull/7414#issuecomment-1692344247). It is probably better to let the users choose by themselves.

henon commented on src/MudBlazor/Components/Stepper/MudStepperStep.razor:
```diff
@@ -0,0 +1,96 @@
+    public RenderFragment ChildContent { get; set; }
+
+    [Parameter]
+    public string Title { get; set; }
```
I'd suggest adding a render fragment for the step title which is only used if the `Title` parameter is `null`. Or a `RenderFragment<string> TitleTemplate` which is used to render the title if set. For instance for MudTab we had a `string Title` param which was not enough for users, eventually we had to add a render fragment alternative. 

Arguably the same logic applies to `SecondaryText` and to the the round stepper icon which shows either a number or the check mark. If we allow an optional RenderFragment<MudSteperStep> StepIconTemplate then we won't need a lot of customization parameters for it in the future.

henon replied:
If we go for templates with parameters instead of simple render fragments the templates should of course be in `MudStepper` instead of `MudStepperStep`.

fondraco replied:
I'm not sure if I completely follow. Should I send there as the parameter the whole stepper? I'm surely down for adding the freedom to template the stepper completely.

henon replied:
> the whole stepper?

Yes, it would allow the template to access everything it needs, right?

fondraco replied:
Yeah, although that would make it necesary to iterate through the List of Steps to find the right one.

henon replied:
I get it, yeah, of course step is better than stepper in that case.

henon commented on src/MudBlazor/Components/Stepper/MudStepperStep.razor:
```diff
@@ -0,0 +1,96 @@
+    public bool Disabled { get; set; }
+
+    /// <summary>
+    /// Raised when tab is clicked TODO: Probably should be called OnActivated or OnCLicked
```
`OnClick` is perfect, it is consistent with naming in other components.

fondraco committed:
MudStepper: Fixed visuals for nonlinear vertical stepper, added hover effect for nonlinear version, added renderfragments for Title and Label content

ScarletKuro commented on src/MudBlazor/Components/Stepper/MudStepper.razor:
```diff
@@ -0,0 +1,123 @@
+        }
+    </div>
+
+    @if (Vertical == false && ActiveStep is not null)
```
nit: `!Vertical`

ScarletKuro commented on src/MudBlazor/Components/Stepper/MudStepper.razor.cs:
```diff
@@ -0,0 +1,251 @@
+    [Parameter]
+    public int ActiveIndex { get; set; }
+
+    [Parameter] public EventCallback<int> ActiveIndexChanged { get; set; }
```
nit: new line after attribute

ScarletKuro commented on src/MudBlazor/Components/Stepper/MudStepper.razor.cs:
```diff
@@ -0,0 +1,251 @@
+    [Parameter]
+    public string NavClass { get; set; }
+
+    [Parameter] public bool NonLinear { get; set; }
```
nit: new line after attribute

ScarletKuro commented on src/MudBlazor/Components/Stepper/MudStepper.razor.cs:
```diff
@@ -0,0 +1,251 @@
+    [Parameter]
+    public bool AlternateLabel { get; set; }
+
+    #region Funcs
```
we have abandoned the practice of using `#region`

ScarletKuro commented on src/MudBlazor/Components/Stepper/MudStepperStep.razor:
```diff
@@ -0,0 +1,102 @@
+@implements IAsyncDisposable
+
+@code {
+    private Boolean _disposed;
```
nit: `Boolean` -> `bool`

ScarletKuro commented on src/MudBlazor.Docs/Pages/Components/Stepper/Examples/StepperDynamicExample.razor:
```diff
@@ -0,0 +1,42 @@
+
+
+@code {
+    List<MudStepperStep> _steps = new List<MudStepperStep>() { new MudStepperStep(){Title = "First"}, new MudStepperStep(){Title = "Second"} };
```
I think the formatting is not correct, and any reformat tool would add space between 
`(){` -> `() {`
 `{Property}` -> `{ Property }`

Int32Overflow commented:
What is the current state?

fondraco commented:
> What is the current state?

I'm working on this in my spare time, so it may take a while. Do you have any suggestions for this issue for me to incorporate into the final component @Int32Overflow ?

fondraco committed:
MudStepper: Disabled steps no longer can be triggered, supporting overrides with colors, error colors work as well, WIP (#7414)

henon commented:
I looked at the examples in the docs and I already like what I see. 

Of course currently the example headlines and descriptions in the docs are still missing. We'll need good section headings and descriptions here, but I guess you know that.

All the examples have a reset button to bring the example back into its initial state which is only visible once you completed all the steps. While this is important for the example, I guess not everyone would want that in their application. So we need an option to disable it (`EnableReset`)? Or maybe the reset button should be entirely removed from the stepper in favor of a separate button in the example that calls a public `Reset()` function? I am undecided about that. At first I didn't know that there is a reset button at the end of the progress and I wished to reset the example but couldn't. So maybe having the reset button be displayed all the time is better.

How can the stepper buttons be configured (i.e. what if I want to define my own buttons with icons, etc., or just change the descriptions of Next, Skip and Previous). For instance, in the last step I'd prefer a "Complete" button instead of a "Next" button, but I guess there can be valid use-cases for custom button descriptions in any step, like "< My Data" to go back and edit my data and "Save >" to complete. 

Minor Issue: In RTL mode the spacing is not right. This can be fixed by using margin start / end classes instead of margin left/right classes.
![image](https://github.com/MudBlazor/MudBlazor/assets/44090/fde0a8dc-6652-47c1-83ef-bf8270e247d1)
The vertical Stepper looks totally destroyed in RTL.

On 100% (no zoom) the number looks uncentered in chrome:
![image](https://github.com/MudBlazor/MudBlazor/assets/44090/2e9e456c-c9e4-46f1-9bfc-b351665e3af6)
Strangely when zooming I can find zoom levels where it is perfectly centered, so I don't know if that is a Chrome render bug or not.
110% looks good for example
![image](https://github.com/MudBlazor/MudBlazor/assets/44090/7ab603b6-6731-4ef1-9faf-3b6e1fa522f4)

How are we on customization with RenderFragments? Can the step title, and the step icon be entirely replaced with render fragments? Maybe also the action bar with the buttons so you can make your own prev/next buttons (for that we need an api that can be called from those custom buttons) 

Hope the feedback helps. All in all, great job so far! I am sure @mckaragoz is gonna have some feedback as well.

mckaragoz commented:
I examine the component, thanks for your efforts, especially vertical mode looks awesome. Have a few things (compared with extension one):

1. Need localizable strings. "Next" "Previous" are hardcoded to English.
2. Currently mobile view is not supported. Some of the steppers has visual issues on mobile. Need a MobileView parameter for a different design i think.
3. I prefer `MudStep` name instead of `MudStepperStep`
4. Component is currently not customizable. Its hardcoded to `Color.Primary` and `Variant.Text`. We would add these parameters for whole component.
5. We can add RenderFragment into action buttons section. User may need to override.
6. We can add a Result Step (optional) that shows a content when all steps completed.
7. When next step is disabled, next button also disabled. Im not sure, but sounds not the best behavior.
8. Skipped steps has no indicator, looks the same with untouched steps.

fondraco commented:
I'll prepare a list of tasks to do from all your observations, and I'll update it continuously.
- [ ] Prepare examples, headlines and descriptions in the docs
- [ ] Reset button should be visible only if the param is set -> `Resettable`
- [ ] Create a public `Reset()` function
- [ ] Fix RTL on the horizontal Stepper
- [ ] Fix RTL on the vertical Stepper
- [ ] Fix the uncentered number in the step circle
- [ ] Add RenderFragment for the Stepper Label (Title, subtitle and the circle)
- [ ] Add RenderFragment for the controls
- [ ] Add localization
- [ ] Create a mobile-friendly design
- [ ] Add RenderFragment to render a result page
- [ ] Label steps that have been skipped
- [ ] Rename `MudStepperStep` to `MudStep`

To address other points previously mentioned not included in the tasks above or are, but in a simplified form.


> How can the stepper buttons be configured (i.e. what if I want to define my own buttons with icons, etc., or just change the descriptions of Next, Skip and Previous). For instance, in the last step I'd prefer a "Complete" button instead of a "Next" button, but I guess there can be valid use-cases for custom button descriptions in any step, like "< My Data" to go back and edit my data and "Save >" to complete.

I'd maybe go for a slightly different approach here. I'd suggest to create a component `MudStepperActions` that can be added either as the child of the `MudStepper` and rendered inside or added outside and be binded with the main component. The component itself would have the options to override its design and implement via RenderFragment its own actions with all the buttons custom. This would give the user new ways of implementing the actions and put them outside of the `MudStepper`. I'm quite hesitant about this approach, because it's not common in this library and might create a lot of confusion.

> 4\. Component is currently not customizable. Its hardcoded to `Color.Primary` and `Variant.Text`. We would add these parameters for whole component.

The Colors are actually customizable, there is the CurrentColor, ErrorColor for that. It can be customized on the global level or on each step

> 7\. When next step is disabled, next button also disabled. Im not sure, but sounds not the best behavior.

If the Stepper is `Linear`, the button should be disabled. If the Stepper is nonlinear, then the button should be enabled.

fondraco committed:
MudStepper: Renamed MudStepperStep to MudStep, Fixed RTL (#7414)

fondraco committed:
MudStepper: Added localization (#7414)

fondraco committed:
MudStepper: Created render fragments for the title and number circle, Added ConnectorTemplate and CompletedContent (#7414)

fondraco committed:
MudStepper: Updated for mobile resolution, prepared examples, polished features (#7414)

fondraco commented:
Stepper can now be customized via `TitleTemplate`, `LabelTemplate` and `ConnectorTemplate`

https://github.com/MudBlazor/MudBlazor/assets/49674671/2febe5a5-2ee1-4e46-af4f-d4d8417ec364

It also supports dynamic steps and becomes scrollable if there are way too many of them. This behaviour can be turned off but it's true by default.

https://github.com/MudBlazor/MudBlazor/assets/49674671/98cd2006-618f-462e-952e-82d9f55200dd

henon committed:
Minor fixes/changes to the Stepper Binding example

henon commented:
I am in the middle of reviewing, not completed yet. Looking good so far. I made some minor changes to the Stepper Binding Example so that it works better and pushed them to your branch.

henon commented on src/MudBlazor/Components/Stepper/MudStepper.razor:
```diff
@@ -0,0 +1,147 @@
+                .AddClass("mud-stepper-nav-step-disabled", step.Disabled)
+                .Build();
+
+            @if (NonLinear)
+            {
+                <div class="@stepClassname" role="tab" aria-controls="@step.Title" aria-selected="@(isActive ? "true" : "false")"
+                     @onclick="@((e) => ProcessStep(step, e, StepInteractionType.Activate))">
+                    @RenderStepLabel(step, isActive)
+                </div>
+            }
+            else
+            {
+                <div class="@stepClassname" role="tab" aria-controls="@step.Title" aria-selected="@(isActive ? "true" : "false")">
+                    @RenderStepLabel(step, isActive)
+                </div>
+            }
```
Could we avoid this duplication by just doing nothing in the `@onclick` handler if `NonLinar` is true?

fondraco replied:
Oh, yeah, sure. Didn't think about that one.

henon commented:
About `NonLinear` steppers: Right now it is not possible to complete a `NonLinear stepper if you did not visit all non-skippable steps. You will always be stuck with a Next button at the end and clicking it does nothing. I think that the nature of a `NonLinear` stepper is that the user does not have to visit all steps, otherwise you would go with a progressive one. Also, for a non-linear stepper non-skippable steps make no sense or you would not allow the user to jump all to the end. 

So I think this logic here: `public bool IsCompleted => _steps.Any() && _steps.Where(x => !x.Skippable).All(x => x.Completed);` 

which determines if the user can visit the `CompletedContent` should only be applied for progressive steppers, and for non-linear ones we just return `true`, which means the user can always go to the `CompletedContent` from the last step no matter what steps they visted or not. If necessary, any validation can be done in the completed content by the user, we don't have to pull that complexity into our non-linear stepper. 

So I suggest changing it to `public bool IsCompleted => NonLinear || _steps.Any() && _steps.Where(x => !x.Skippable).All(x => x.Completed);`

A thing we might want to add for progressive steppers: Being able to jump back to visited steps by clicking the header. This would allow to jump around in the visited steps but not jump ahead as is the case with non-linear steppers. We can add that in a later PR.

henon commented:
Oh by the way, it would be awesome to make the Next-Button on the final step read `Complete` instead of `Next`

henon commented:
As a user of the stepper, how would you lock the `Next` button until the user completed the step (i.e. by fully filling yout a form)?

henon commented:
I found the dynamic example crashes if you remove all the steps and then try to add one

```
blazor.server.js:1 [2023-11-24T10:41:21.060Z] Error: System.ArgumentOutOfRangeException: Index was out of range. Must be non-negative and less than the size of the collection. (Parameter 'index')
   at System.Collections.Generic.List`1.get_Item(Int32 index)
   at MudBlazor.MudStepper.get_CanGoToNextStep() in D:\MudBlazor.git\src\MudBlazor\Components\Stepper\MudStepper.razor.cs:line 107
   at MudBlazor.MudStepper.<BuildRenderTree>b__112_1(RenderTreeBuilder __builder2) in D:\MudBlazor.git\src\MudBlazor\Components\Stepper\MudStepper.razor:line 77
   at Microsoft.AspNetCore.Components.Rendering.RenderTreeBuilder.AddContent(Int32 sequence, RenderFragment fragment)
   at MudBlazor.MudCardActions.BuildRenderTree(RenderTreeBuilder __builder) in D:\MudBlazor.git\src\MudBlazor\Components\Card\MudCardActions.razor:line 5
   at Microsoft.AspNetCore.Components.Rendering.ComponentState.RenderIntoBatch(RenderBatchBuilder batchBuilder, RenderFragment renderFragment, Exception& renderFragmentException)
```

henon committed:
MudStep: make HasError update parent when changed and add example

henon commented:
I pushed another fix. The step's `HasError` state did not update the parent when changed. I added another switch to the binding example showing off the error state.

![image](https://github.com/MudBlazor/MudBlazor/assets/44090/003044c9-b80a-4547-80c5-f20b708b3295)

fondraco committed:
MudStepper: Removed code duplication ($7414)

henon committed:
MudStepper: Add HasError to example subscription

henon committed:
typo

henon committed:
Merge remote-tracking branch '_MudBlazor/dev' into feature/stepper

henon commented:
Merged dev to see if that will fix the failing build

henon commented:
![image](https://github.com/MudBlazor/MudBlazor/assets/44090/d736c6b1-2ddc-49e1-8dd2-e9e2c90d451d)

I am not sure about the naming of `AlternateLabel`. Maybe `CenterLabels`?

henon commented:
Aha, the tests fail because of missing parameter categories

```
Failed AllComponentPropertiesHaveCategories [28 ms]
  Error Message:
     Some component properties don't have categories.
```

henon commented:
The functionality of the dynamic example is clear but I do not understand what the explanation is about here:

![image](https://github.com/MudBlazor/MudBlazor/assets/44090/24cf4dc9-5645-4b7a-bfe0-7e65f8c0f6af)

henon commented:
I suggest we remove the last two examples and combine them into the non-linear stepper example, so we have one example that lets the users switch between progressive and non-linear as well as horizontal and vertial. The non-linear example could show both non-linearity and vertical layout by default.

henon committed:
MudStepper: improve a doc comment

henon committed:
Stepper docs: minor improvements

henon committed:
Stepper docs: fix dynamic example description

henon committed:
MudStep: add IsActive property

henon committed:
Stepper docs: pimped the customization example a bit

henon commented:
Nice customizability, a screenshot for the record:
![image](https://github.com/MudBlazor/MudBlazor/assets/44090/664ccab9-6c0e-417b-bcf3-7e6bb5741810)

ScarletKuro commented on src/MudBlazor.Docs/Pages/Components/Stepper/Examples/StepperBindedStepExample.razor:
```diff
@@ -0,0 +1,36 @@
+<MudNumericField @bind-Value="_index" Label="Current index" Variant="Variant.Text" Min="0" Max="5" />
+
+
+@code {
```
redundant code block since we already have one below?

henon replied:
I removed it.

ScarletKuro commented:
It looks good, tho not all public methods follow our new guideline which is to have "Async" suffix if it's a Task method.

henon committed:
MudStepper: fix potential nullref exception

henon committed:
Docs: remove redundant code block

henon commented:
@ScarletKuro We might also want to turn on nullable reference types in stepper as per our new standard

henon committed:
minor change

henon reviewed (changes requested):
I love it. Code is clear and concise, customizability is awesome and easy. Usage is very easy. 

There are a few minor things to fix as pointed out above. My suggestion is that the following important things are addressed before we release it as experimental:
- [ ] Allowing users to lock the Next button until the task of a step is completed, because this is essential functionality that probably everyone will need
- [ ] Customization of the action panel, maybe just a RF for the entire action panel plus an api for programmatically calling Previous, Skip and Next? 
- [ ] Remove the reset button and add a manual reset button to the examples instead so people playing with them can set the stepper back.
- [ ] at least 80% test coverage

henon committed:
Docs: add the experimental notice

henon committed:
move stepper css from _chip.scss to _stepper.scss

henon committed:
fix warnings which are now errors

henon commented:
@fondraco I rebased and fixed conflicts and warnings

henon committed:
Add a test and fix param category and user attributes requirements

fondraco commented:
> @fondraco I rebased and fixed conflicts and warnings

Thank you, I'm finishing a project in more or less two weeks. Then I'll complete the Stepper. I know it's been a while hahha.

ScarletKuro commented:
> Thank you, I'm finishing a project in more or less two weeks. Then I'll complete the Stepper. I know it's been a while hahha.

This would be great, looking forward for this component.
https://github.com/MudBlazor/MudBlazor/blob/d81b2c6135348b439333b6befa34418aeaf652b1/src/MudBlazor/Components/Stepper/MudStep.razor#L114-L120
You probably can optimize this now with our new `ParameterState` framework https://github.com/MudBlazor/MudBlazor/pull/8258
Also you'd be able to avoid writing to the components parameters:
https://github.com/MudBlazor/MudBlazor/blob/d81b2c6135348b439333b6befa34418aeaf652b1/src/MudBlazor/Components/Stepper/MudStep.razor#L139
Which is not recommended by [Microsoft ](https://learn.microsoft.com/en-us/aspnet/core/blazor/components/overwriting-parameters?view=aspnetcore-8.0)
You'd have to call the `SetValueAsync` method (it would also call the EventCallback if any) and use `Value` from state variable, if you need to read the variable for example here:
https://github.com/MudBlazor/MudBlazor/blob/d81b2c6135348b439333b6befa34418aeaf652b1/src/MudBlazor/Components/Stepper/MudStepper.razor#L90

henon committed:
Stepper: a primitive functional test

henon commented:
yes, but I want to create full test coverage before refactoring towards using our new ParameterState framework

henon committed:
More tests

ScarletKuro commented on src/MudBlazor/Components/Stepper/MudStepper.razor.cs:
```diff
@@ -0,0 +1,281 @@
+    private int _activeIndex = -1;
+    private HashSet<MudStep> _skippedSteps = new();
+
+    public IReadOnlyList<MudStep> Steps { get; private set; }
```
I see that the `Steps` are not set anymore anywhere so the `private set` is redundant.
But i see that the `_steps` is getting modified this means that the `Steps` and `_steps` won't be in sync once _steps are modified.
Since you set the `Steps` is constructor it will be evaluated only once, you need to make `Steps => _steps` then it will re-evaluate every time the property is called.

ScarletKuro commented on src/MudBlazor/Components/Stepper/MudStepper.razor.cs:
```diff
@@ -0,0 +1,281 @@
+        ProcessStep(_steps[0], new MouseEventArgs(), StepInteractionType.Activate);
+    }
+
+    internal void Refresh() => StateHasChanged();
```
We don't use that anymore as `MudCompoentBase` implements `IMudStateHasChanged`, if you need to call the `StateHasChanged` from outside, you just cast the component to that interface.

henon committed:
Changes requested by Kuro

henon commented:
@fondraco I may have found a potential bug:

If a step is Completed and Error then the error won't show. Is it intentional or do you think as well, that `Error` should have higher priority than `Completed`.

![image](https://github.com/MudBlazor/MudBlazor/assets/44090/75e39276-3753-4380-ba27-6a789c3dd64a)

I haven't checked how `Disabled` combined with `Completed` and `Error` behaves, but I'll write a unit test for this and I'd like to know what is the expected and correct behavior.

ScarletKuro commented on src/MudBlazor/Components/Stepper/MudStepper.razor.cs:
```diff
@@ -17,21 +17,21 @@ public partial class MudStepper : MudComponentBase
     private HashSet<MudStep> _skippedSteps = new();
 
-    public IReadOnlyList<MudStep> Steps { get; private set; }
+    internal IReadOnlyList<MudStep> Steps => _steps;
```
Should we leave this public considering we have an analogue request for `MudExpansionPanels` https://github.com/MudBlazor/MudBlazor/issues/7756 that has currently private _panels?

ScarletKuro replied:
Off Topic: I wonder if it would make sense for components that holds collection of items to implement `IEnumerable` / `IReadOnlyCollection` / `ReadOnlyList` (if we can get the count) directly on the component? I can see use in this.

henon replied:
I don't know what it can be used for. The steps are accessible through razor anyway.

henon replied:
Hmm, you may be right, the steps have public API such as SetHasError, SetCompleted and SetDisabled.

henon committed:
Add more tests and fixes/improvements found while testing

henon committed:
Rename to use Async postfix and some other renaming

henon committed:
More tests and related fixes

henon committed:
CompleteContent test

henon committed:
revert my own mistake

henon committed:
MudStep.razor => MudStep.cs

henon committed:
Merge commit '9e39a7d07dd8d99da32ef26c4a0809e91cdb2e2d' into feature/stepper

henon committed:
MudStep: refactor to use State framework

henon committed:
MudStep: enable nullable

henon committed:
MudStepper: refactor to use state framework

srs-adamr commented:
Has this been abandoned?

henon commented:
No, this will come after v7.0.0 is released. It is almost done

dennisrahmen commented:
> If a step is Completed and Error then the error won't show.

What happens when I completed a step, then go back and clear a required field.
Would that still show completed?

Because then the user would not know at which step they made a mistake.

henon commented:
@dennisrahmen Yes, that's why I believe this should be changed. Error should override Completed.

dennisrahmen commented:
@henon I believe that's the better solution.

henon commented:
One of the missing pieces of this PR is a freely configurable action bar where the user can place their own previous-, next- and skip buttons as well as a reset button if they want. There are two possibilities I can come up with:

1) Using a references to the stepper and calling public API such as `Previous()`, `Next()` etc. Usage would look like this:

```html
<MudStepper @ref="stepper">
  ...
  <ActionContent>
    <MudButton Click="@(stepper.Previous())"> ... </MudButton>
    <MudButton Click="@(stepper.Skip())"> ... </MudButton>
    <MudButton Click="@(stepper.Next())"> ... </MudButton>
    <MudButton Click="@(stepper.Reset())"> ... </MudButton>
  </ActionContent>
</MudStepper>
```

2) Extend our `IActivatable` interface with an optional action argument so we can distinguish between different actions. 

```c#
    public interface IActivatable
    {
        void Activate(object activator, MouseEventArgs args, string action=null);
    }
```

Every button would be placed in an `Activator` implementing `IActivatable` itself. The activators listen for an activation in their content and transmit the activation along with their configured `Action` they receive up to the stepper which also implements `IActivatable`. Usage would look like this:

```html
<MudStepper>
  ...
  <ActionContent>
    <Activator Action="@StepperActions.Previous">
      <MudButton> ... </MudButton>
    </Activator>
    <Activator Action="@StepperActions.Skip">
      <MudButton> ... </MudButton>
    </Activator>
    <Activator Action="@StepperActions.Next">
      <MudButton> ... </MudButton>
    </Activator>
    <Activator Action="@StepperActions.Reset">
      <MudButton> ... </MudButton>
    </Activator>
  </ActionContent>
</MudStepper>
```

I tend towards the first solution because it is simpler, but the second solution would work without needing a reference to the stepper and would potentially allow passing activations up multiple component levels without having to know the receiver so it has a lot more potential and it can also be used in other places such as menus, datagrid, table (i.e. for custom paginators) etc.

For instance, recently we had an issue with `MudMenu` recently where the user wanted to place multiple buttons in the menu but only one of them should activate the Menu's close action. One way to achieve this would be to make `<MudMenu>` listen only to a specific action (i.e. `MenuAction.Close`)  and you have to use an `<Activator>`  (or  `<MudActivator>` ?? ) to close it:

```html
<MudMenu>
  <MudButton>This button doesn't close</MudButton>
  <MudActivator Action="@MenuActions.Close">
    <MudButton>Closes</MudButton>
  </MudActivator>
</MudMenu>
```

Either that, or we turn it around so that placing a button activates the close like it is now and placing a button that doesn't activate the menu is done like this

```html
<MudMenu>
  <MudActivator Enabled="false">
    <MudButton>This button doesn't close</MudButton>
  </MudActivator>
  <MudButton>Closes</MudButton>
</MudMenu>
```

henon commented:
As a second thought, I also thought about adding an `Action` parameter to `MudButton`, `MudIconButton` etc then the `MudActivator` wouldn't be needed any more. But I am very unsure about this. The Usage would look nice though:

```html
<MudStepper @ref="stepper">
  ...
  <ActionContent>
    <MudButton Action="@StepperActions.Previous"> ... </MudButton>
    <MudButton Action="@StepperActions.Skip"> ... </MudButton>
    <MudButton Action="@StepperActions.Next"> ... </MudButton>
    <MudButton Action="@StepperActions.Reset"> ... </MudButton>
  </ActionContent>
</MudStepper>
```

ScarletKuro commented:
What about this

Component:
```razor
<h3>MudStepper</h3>

@if (ActionContent is not null)
{
    @ActionContent(new StepperButtonContext(this))
}

@code {
    [Parameter]
    public RenderFragment<StepperButtonContext>? ActionContent { get; set; }

    internal Task SkipAsync()
    {
        Console.WriteLine(nameof(SkipAsync));
        return Task.CompletedTask;
    }

    internal Task NextAsync()
    {
        Console.WriteLine(nameof(NextAsync));
        return Task.CompletedTask;
    }

    internal Task PreviousAsync()
    {
        Console.WriteLine(nameof(PreviousAsync));
        return Task.CompletedTask;
    }

    internal Task ResetAsync()
    {
        Console.WriteLine(nameof(ResetAsync));
        return Task.CompletedTask;
    }
}
```
Context:
```C#
public class StepperButtonContext
{
    private readonly MudStepper _mudStepper;

    public StepperButtonContext(MudStepper mudStepper)
    {
        _mudStepper = mudStepper;
    }

    public Task SkipAsync() => _mudStepper.SkipAsync();

    public Task NextAsync() => _mudStepper.NextAsync();

    public Task PreviousAsync() => _mudStepper.PreviousAsync();

    public Task ResetAsync() => _mudStepper.ResetAsync();
}
```

Usage:
```razor
<MudStepper>
    <ActionContent Context="stepper">
        <button class="btn btn-primary" @onclick="@stepper.NextAsync">Next</button>
        <button class="btn btn-primary" @onclick="@stepper.PreviousAsync">Previous</button>
    </ActionContent>
</MudStepper>
```

I tried and it works.

henon commented:
The context is of course the best and simplest solution. Thanks for opening my eyes @ScarletKuro ;)

ScarletKuro commented:
It will be also possible to intercept:
```razor
<MudStepper>
    <ActionContent Context="stepper">
        <button class="btn btn-primary" @onclick="() => OwnMethod(stepper)">Next</button>
    </ActionContent>
</MudStepper>


@code {
    private Task OwnMethod(StepperButtonContext context)
    {
	MethodBefore();
        await context.NextAsync();
        MethodAfter();
    }
}
```

srs-adamr commented:
Next release? 🙏

PlayerModu commented:
In this component can the stepper action buttons be placed at the top and/or bottom of the stepper? Or are they fixed at the bottom? I mean like Back/Next

henon commented:
By default at the bottom. But you can always put your own buttons somewhere else and just call the public `NextStep()` and `PreviousStep()` API.

henon commented:
I continued this PR in #9911

We now have the possibility for custom action buttons. This is the last major missing piece. I think we can merge it as experimental like this. Only minor changes are to be expected.

![image](https://github.com/user-attachments/assets/ec216faf-94ff-436e-97f3-f51217ab94f0)

Will add some more tests, then merge.
